File: psana_examples/trunk/src/dump_imp.py
#--------------------------------------------------------------------------
# File and Version Information:
#  $Id$
#
# Description:
#  Psana user analysis module dump_imp...
#
#------------------------------------------------------------------------

"""Example module for psana which dumps Imp data.

This software was developed for the LCLS project.  If you use all or 
part of it, please give an appropriate acknowledgment.

@version $Id$

@author Andy Salnikov
"""

#------------------------------
#  Module's version from SVN --
#------------------------------
__version__ = "$Revision$"
# $Source$

#--------------------------------
#  Imports of standard modules --
#--------------------------------
import sys
import logging

#-----------------------------
# Imports for other modules --
#-----------------------------
from psana import *

#----------------------------------
# Local non-exported definitions --
#----------------------------------

#---------------------
#  Class definition --
#---------------------
class dump_imp (object) :
    '''Class whose instance will be used as a user analysis module.'''


    #----------------
    #  Constructor --
    #----------------
    def __init__(self):
        self.m_src = self.configSrc('source', ':Imp')

    #-------------------
    #  Public methods --
    #-------------------

    def beginRun( self, evt, env ) :

        config = env.configStore().get(Imp.Config, self.m_src)
        if config:
            print "dump_imp: %s: %s" % (config.__class__.__name__, self.m_src)
            print "  range =", config.range()
            print "  calRange =", config.calRange()
            print "  reset =", config.reset()
            print "  biasData =", config.biasData()
            print "  calData =", config.calData()
            print "  biasDacData =", config.biasDacData()
            print "  calStrobe =", config.calStrobe()
            print "  numberOfSamples =", config.numberOfSamples()
            print "  trigDelay =", config.trigDelay()
            print "  adcDelay =", config.adcDelay()

    def event( self, evt, env ) :

        data = evt.get(Imp.Element, self.m_src)
        if not data:
            return

        print "dump_imp: %s: %s" % (data.__class__.__name__, self.m_src)

        print "  vc =", data.vc()
        print "  lane =", data.lane()
        print "  frameNumber =", data.frameNumber()
        print "  range =", data.range()

        laneStatus = data.laneStatus()
        print "  laneStatus.linkErrCount =", laneStatus.linkErrCount()
        print "  laneStatus.linkDownCount =", laneStatus.linkDownCount()
        print "  laneStatus.cellErrCount =", laneStatus.cellErrCount()
        print "  laneStatus.rxCount =", laneStatus.rxCount()
        print "  laneStatus.locLinked =", laneStatus.locLinked()
        print "  laneStatus.remLinked =", laneStatus.remLinked()
        print "  laneStatus.zeros =", laneStatus.zeros()
        print "  laneStatus.powersOkay =", laneStatus.powersOkay()
        
        for i, sample in enumerate(data.samples()):
            print "  sample[%d]: channels = %s" % (i, sample.channels())

File: psana_examples/tags/V00-05-00/src/DumpImp.cpp
//--------------------------------------------------------------------------
// File and Version Information:
// 	$Id$
//
// Description:
//	Class DumpImp...
//
// Author List:
//      Andy Salnikov
//
//------------------------------------------------------------------------

//-----------------------
// This Class's Header --
//-----------------------
#include "psana_examples/DumpImp.h"

//-----------------
// C/C++ Headers --
//-----------------

//-------------------------------
// Collaborating Class Headers --
//-------------------------------
#include "MsgLogger/MsgLogger.h"
#include "psddl_psana/imp.ddl.h"

//-----------------------------------------------------------------------
// Local Macros, Typedefs, Structures, Unions and Forward Declarations --
//-----------------------------------------------------------------------

// This declares this class as psana module
using namespace psana_examples;
PSANA_MODULE_FACTORY(DumpImp)

//		----------------------------------------
// 		-- Public Function Member Definitions --
//		----------------------------------------

namespace psana_examples {

//----------------
// Constructors --
//----------------
DumpImp::DumpImp (const std::string& name)
  : Module(name)
  , m_src()
{
  m_src = configSrc("source", "DetInfo(:Imp)");
}

//--------------
// Destructor --
//--------------
DumpImp::~DumpImp ()
{
}

/// Method which is called at the beginning of the run
void 
DumpImp::beginRun(Event& evt, Env& env)
{
  MsgLog(name(), info, "in beginRun()");

  Pds::Src actualSrc;
  shared_ptr<Psana::Imp::ConfigV1> config1 = env.configStore().get(m_src, &actualSrc);
  if (config1) {
    WithMsgLog(name(), info, str) {
      str << "Imp::ConfigV1 at " << actualSrc;
      str << "\n  range = " << config1->range();
      str << "\n  calRange = " << config1->calRange();
      str << "\n  reset = " << config1->reset();
      str << "\n  biasData = " << config1->biasData();
      str << "\n  calData = " << config1->calData();
      str << "\n  biasDacData = " << config1->biasDacData();
      str << "\n  calStrobe = " << config1->calStrobe();
      str << "\n  numberOfSamples = " << config1->numberOfSamples();
      str << "\n  trigDelay = " << config1->trigDelay();
      str << "\n  adcDelay = " << config1->adcDelay();
    }
  }
}

/// Method which is called with event data, this is the only required 
/// method, all other methods are optional
void 
DumpImp::event(Event& evt, Env& env)
{
  Pds::Src actualSrc;
  shared_ptr<Psana::Imp::ElementV1> data1 = evt.get(m_src, "", &actualSrc);
  if (data1) {

    WithMsgLog(name(), info, str) {
      str << "Imp::ElementV1 at " << actualSrc
          << "\n  vc = " << int(data1->vc())
          << "\n  lane = " << int(data1->lane())
          << "\n  frameNumber = " << data1->frameNumber()
          << "\n  range = " << data1->range();

      Psana::Imp::LaneStatus laneStatus = data1->laneStatus();
      str << "\n  laneStatus.linkErrCount = " << int(laneStatus.linkErrCount())
          << "\n  laneStatus.linkDownCount = " << int(laneStatus.linkDownCount())
          << "\n  laneStatus.cellErrCount = " << int(laneStatus.cellErrCount())
          << "\n  laneStatus.rxCount = " << int(laneStatus.rxCount())
          << "\n  laneStatus.locLinked = " << int(laneStatus.locLinked())
          << "\n  laneStatus.remLinked = " << int(laneStatus.remLinked())
          << "\n  laneStatus.zeros = " << int(laneStatus.zeros())
          << "\n  laneStatus.powersOkay = " << int(laneStatus.powersOkay());

      // get samples 1-d array
      const ndarray<const Psana::Imp::Sample, 1>& samples = data1->samples();
      for (unsigned i = 0; i != samples.size(); ++ i) {
        str << "\n  sample[" << i << "]: channels = " << samples[i].channels();
      }

    }
  }
}
  
} // namespace psana_examples
